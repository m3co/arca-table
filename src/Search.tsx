import * as React from 'react';
import { Combobox, Row, SearchMethod, Params } from 'arca-redux';
import { v4 as uuid4 } from 'uuid';

interface Props {
  Row: Row;
  Setup: Combobox;
  list?: string;
  value: string;
  search: SearchMethod;
  onBlur?: (currentValue: string) => void;
  onEnter?: (currentValue: string) => void;
  onEsc?: () => void;
}

interface RowOption {
  label: string | number;
  value: string | number;
}

interface State {
  Rows: RowOption[];
  value: string;
}

export class Search
  extends React.Component<Props, State>
{
  public static defaultProps = {
    list: uuid4(),
  }

  public state: State = {
    Rows: [],
    value: this.props.value,
  }

  private columns = Object.keys(this.props.Row) as (keyof Row)[];
  private params = Object.keys(this.props.Setup.Params)
    .reduce((acc: Params, key: string): Params => {
      const column = this.columns
        .find((column): boolean => column === this.props.Setup.Params[key]);
      column && (acc[key] = this.props.Row[column]);
      return acc;
    }, {});

  private search(params: Params): void {
    const { Setup, search } = this.props;
    search(Setup.Source, Setup.Value, params)
      .then((response): void => {
        const Rows = response.Result
          .map((result): RowOption => ({
            value: result[Setup.Value] || '',
            label: result[Setup.Display] || '',
          }));
        this.setState((state: State): State => ({
          ...state,
          Rows,
        }));
      });
  }

  public constructor(props: Props) {
    super(props);
    this.search(this.params);
  }

  private onChange = (event: React.ChangeEvent<HTMLInputElement>): void => {
    const { state, props: { Setup } } = this;
    this.search({
      ...this.params,
      [Setup.Value]: event.target.value,
    });

    this.setState({
      ...state,
      value: event.target.value,
    });
  }

  private onKeyUp = (event: React.KeyboardEvent<HTMLInputElement>): void => {
    const extra = {
      altKey: !!event.altKey,
      ctrlKey: !!event.ctrlKey,
      metaKey: !!event.metaKey,
      shiftKey: !!event.shiftKey,
    };
    if (extra.altKey || extra.ctrlKey || extra.metaKey || extra.shiftKey) {
      return;
    }
    if (event.key == 'Enter' && this.props.onEnter) {
      this.props.onEnter(this.state.value);
    }
    if (event.key == 'Escape' && this.props.onEsc) {
      this.props.onEsc();
    }
  }

  private onBlur = (): void => {
    if (this.props.onBlur) this.props.onBlur(this.state.value);
  }

  public render(): JSX.Element {
    const { state, props: { list } } = this;
    return (
      <React.Fragment>
        <input
          list={list}
          autoFocus={true}
          onChange={this.onChange}
          onKeyUp={this.onKeyUp}
          onBlur={this.onBlur}
          value={state.value} />
        <datalist id={list}>
          {
            state.Rows.map((row, key): JSX.Element =>
              <option
                key={key}
                label={row.label.toString()}
                value={row.value} />
            )
          }
        </datalist>
      </React.Fragment>
    );
  }
}
