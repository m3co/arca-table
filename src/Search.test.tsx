
import * as React from 'react';
import { mount } from 'enzyme';
import Socket from 'socket.io-client';

import { Search } from './Search';
import { Combobox, createSearchSocket,
  State, Fields, Params,
  FACADBuiltInCategories } from 'arca-redux';

interface Response {
  ID: string;
  Method: "Search";
  Context: {
    Field: keyof Fields;
    Source: keyof State["Source"];
  };
  Result: Params[];
  Error: {} | null;
};

const URL = 'http://localhost:8086';

interface MockFns {
  onBlur?: (currentValue: string | number | boolean) => void;
  onEnter?: (currentValue: string | number | boolean) => void;
  onEsc?: () => void;
}

function prepareMock2(searchMock: (Source: keyof State["Source"],
  Field: keyof Fields,
  Params: Params) => Promise<Response>, fns?: MockFns): JSX.Element {

  const combobox: Combobox = {
    Source: "FACAD-BuiltInCategories",
    Display: "BuiltInCategory",
    Value: "BuiltInCategory",
    Params: {
      BuiltInCategory: "BuiltInCategory"
    }
  };
  const row: FACADBuiltInCategories = {
    Row: {
      BuiltInCategory: 'INVALID',
    },
    PK: {
      BuiltInCategory: 'INVALID',
    },
    Raw: null,
    Name: 'FACAD-BuiltInCategories',
  };

  return (<Search
    list={'475e016e-6423-4462-bd91-0535a7e384fc'}
    search={searchMock}
    Row={row.Row}
    Setup={combobox}
    value={'INVALID'}
    onBlur={fns ? fns.onBlur : undefined}
    onEnter={fns ? fns.onEnter : undefined}
    onEsc={fns ? fns.onEsc : undefined} />)
}

test('Search normally renders', (done): void => {
  const io = Socket(URL);
  const search = createSearchSocket(io);
  let transition: Promise<Response>;
  const searchMock = (Source: keyof State["Source"],
    Field: keyof Fields,
    Params: Params): Promise<Response> => {
    const r = search(Source, Field, Params, '5832c751-a6a9-4cd2-9235-157d90cb95d3');
    transition = r;
    return r;
  };

  const el = prepareMock2(searchMock);
  const wrapper = mount(el);
  setTimeout((): void => {
    transition.then((response: Response): void => {
      setTimeout((): void => {
        wrapper.update();
        expect(response).toMatchSnapshot();
        expect(wrapper).toMatchSnapshot();
        done();
      });
    });
  })
});

const searchMock = (Source: keyof State["Source"],
  Field: keyof Fields,
  Params: Params): Promise<Response> => {
  const response: Response = {
    ID: '',
    Method: "Search",
    Context: {
      Field,
      Source,
    },
    Result: [{BuiltInCategory:"OST_InternalLineLoads"},{BuiltInCategory:"OST_InternalLineLoadTags"},{BuiltInCategory:"OST_LineLoads"},{BuiltInCategory:"OST_LineLoadTags"}],
    Error: null,
  };
  return Promise.resolve(response);
};

test('Search changes its value', (): void => {
  const el = prepareMock2(searchMock);
  const wrapper = mount(el);
  wrapper.update();
  wrapper.find('input').simulate('change', { target: { value: 'ANOTHER' } });
  expect(wrapper).toMatchSnapshot();
});

test('Search fires onBlur', (): void => {
  let onBlurReached = false;
  const onBlur = (currentValue: string | number | boolean): void => {
    expect(currentValue).toBe('INVALID');
    onBlurReached = true;
  }

  const el = prepareMock2(searchMock, {onBlur});
  const wrapper = mount(el);
  wrapper.find('input').simulate('blur');
  expect(onBlurReached).toBe(true);
});

test('Search ignores onEnter if altKey', (): void => {
  const value = 'INVALID';
  const onEnter = (currentValue: string | number | boolean): void => {
    expect(currentValue).toBe(value);
    fail();
  }
  const el = prepareMock2(searchMock, {onEnter});
  const wrapper = mount(el);
  wrapper.find('input').simulate('keyup', { key: 'Enter', altKey: true });
});

test('Search accepts onEnter', (): void => {
  const value = 'INVALID';
  const onEnter = (currentValue: string | number | boolean): void => {
    expect(currentValue).toBe(value);
  }
  const el = prepareMock2(searchMock, {onEnter});
  const wrapper = mount(el);
  wrapper.find('input').simulate('keyup', { key: 'Enter' });
});

test('Input fires onEsc', (): void => {
  let onEscReached = false;
  const onEsc = (): void => {
    onEscReached = true;
  }
  const el = prepareMock2(searchMock, {onEsc});
  const wrapper = mount(el);
  wrapper.find('input').simulate('keyup', { key: 'Escape' });
  expect(onEscReached).toBe(true);
});
