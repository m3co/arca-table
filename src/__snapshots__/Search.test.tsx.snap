// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search changes its value 1`] = `
<Search
  Row={
    Object {
      "BuiltInCategory": "INVALID",
    }
  }
  Setup={
    Object {
      "Display": "BuiltInCategory",
      "Params": Object {
        "BuiltInCategory": "BuiltInCategory",
      },
      "Source": "FACAD-BuiltInCategories",
      "Value": "BuiltInCategory",
    }
  }
  list="475e016e-6423-4462-bd91-0535a7e384fc"
  search={[Function]}
  value="INVALID"
>
  <input
    autoFocus={true}
    list="475e016e-6423-4462-bd91-0535a7e384fc"
    onBlur={[Function]}
    onChange={[Function]}
    onKeyUp={[Function]}
    value="ANOTHER"
  />
  <datalist
    id="475e016e-6423-4462-bd91-0535a7e384fc"
  />
</Search>
`;

exports[`Search normally renders 1`] = `
Object {
  "Context": Object {
    "Field": "BuiltInCategory",
    "Source": "FACAD-BuiltInCategories",
  },
  "Error": null,
  "ID": "5832c751-a6a9-4cd2-9235-157d90cb95d3",
  "Method": "Search",
  "Result": Array [
    Object {
      "BuiltInCategory": "OST_InternalLineLoads",
    },
    Object {
      "BuiltInCategory": "OST_InternalLineLoadTags",
    },
    Object {
      "BuiltInCategory": "OST_LineLoads",
    },
    Object {
      "BuiltInCategory": "OST_LineLoadTags",
    },
  ],
}
`;

exports[`Search normally renders 2`] = `
<Search
  Row={
    Object {
      "BuiltInCategory": "INVALID",
    }
  }
  Setup={
    Object {
      "Display": "BuiltInCategory",
      "Params": Object {
        "BuiltInCategory": "BuiltInCategory",
      },
      "Source": "FACAD-BuiltInCategories",
      "Value": "BuiltInCategory",
    }
  }
  list="475e016e-6423-4462-bd91-0535a7e384fc"
  search={[Function]}
  value="INVALID"
>
  <input
    autoFocus={true}
    list="475e016e-6423-4462-bd91-0535a7e384fc"
    onBlur={[Function]}
    onChange={[Function]}
    onKeyUp={[Function]}
    value="INVALID"
  />
  <datalist
    id="475e016e-6423-4462-bd91-0535a7e384fc"
  >
    <option
      key="0"
      label="OST_InternalLineLoads"
      value="OST_InternalLineLoads"
    />
    <option
      key="1"
      label="OST_InternalLineLoadTags"
      value="OST_InternalLineLoadTags"
    />
    <option
      key="2"
      label="OST_LineLoads"
      value="OST_LineLoads"
    />
    <option
      key="3"
      label="OST_LineLoadTags"
      value="OST_LineLoadTags"
    />
  </datalist>
</Search>
`;
